homeassistant/components/music_assistant/media_player.py

```diff
@@ -0,0 +1,751 @@
+        """Send pause command to device."""
+        await self.mass.players.player_command_pause(self.player_id)
+
+    @catch_musicassistant_error
+    async def async_media_stop(self) -> None:
+        """Send stop command to device."""
+        await self.mass.players.player_command_stop(self.player_id)
+
+    @catch_musicassistant_error
+    async def async_media_next_track(self) -> None:
+        """Send next track command to device."""
+        if TYPE_CHECKING:
+            assert self.player.active_source is not None
+        if queue := self.mass.player_queues.get(self.player.active_source):
+            await self.mass.player_queues.queue_command_next(queue.queue_id)
+
+    @catch_musicassistant_error
+    async def async_media_previous_track(self) -> None:
+        """Send previous track command to device."""
+        if TYPE_CHECKING:
+            assert self.player.active_source is not None
+        if queue := self.mass.player_queues.get(self.player.active_source):
+            await self.mass.player_queues.queue_command_previous(queue.queue_id)
+
+    @catch_musicassistant_error
+    async def async_media_seek(self, position: float) -> None:
+        """Send seek command."""
+        position = int(position)
+        if TYPE_CHECKING:
+            assert self.player.active_source is not None
+        if queue := self.mass.player_queues.get(self.player.active_source):
+            await self.mass.player_queues.queue_command_seek(queue.queue_id, position)
+
+    @catch_musicassistant_error
+    async def async_mute_volume(self, mute: bool) -> None:
+        """Mute the volume."""
+        await self.mass.players.player_command_volume_mute(self.player_id, mute)
+
+    @catch_musicassistant_error
+    async def async_set_volume_level(self, volume: float) -> None:
+        """Send new volume_level to device."""
+        volume = int(volume * 100)
+        await self.mass.players.player_command_volume_set(self.player_id, volume)
+
+    @catch_musicassistant_error
+    async def async_volume_up(self) -> None:
+        """Send new volume_level to device."""
+        await self.mass.players.player_command_volume_up(self.player_id)
+
+    @catch_musicassistant_error
+    async def async_volume_down(self) -> None:
+        """Send new volume_level to device."""
+        await self.mass.players.player_command_volume_down(self.player_id)
+
+    @catch_musicassistant_error
+    async def async_turn_on(self) -> None:
+        """Turn on device."""
+        await self.mass.players.player_command_power(self.player_id, True)
+
+    @catch_musicassistant_error
+    async def async_turn_off(self) -> None:
+        """Turn off device."""
+        await self.mass.players.player_command_power(self.player_id, False)
+
+    @catch_musicassistant_error
+    async def async_set_shuffle(self, shuffle: bool) -> None:
+        """Set shuffle state."""
+        if TYPE_CHECKING:
+            assert self.player.active_source is not None
```

Comment:
why would it be None in the first place?

Reply:
active_source is optional in the model

Reply:
But when can it be None?

Reply:
If the player does not (yet) have a source set. In this case its guarenteed to be not None so the type checker needs a little bump

Reply:
Source can be set to a string or None